IBM/carbon-addons-iot-react
# 720 in line editing of data table

Closes #720

**Summary**
_Extend the IoT Data Table to allow for editing of lines row-by-row._

The bulk of the features are implemented outside the actual table as a story with only the minimal required changes to the table and the props:

- New `options.hasRowEdit` prop
- view.toolbar.activeBar prop got new value `rowEdit`
- New `actions.toolbar.onShowRowEdit` callback prop (closing the row edit toolbar is handled by the application)
- New `editDataFunction` prop in the columns objects
- New `i18n.editButtonAria` prop for the table

The table story "Basic table with full rowEdit example" shows a simple but full implementation with different kinds of editable table cells, save/cancel buttons and a toast with an undo-option. 
![image](https://user-images.githubusercontent.com/5167091/81417200-1b9bee00-914b-11ea-9afa-b9b7e9714594.png)

![image](https://user-images.githubusercontent.com/5167091/81405838-bc34e280-9138-11ea-9971-e7e293796308.png)

**Change List (commits, features, bugs, etc)**
- Used explicit prop hasRowEdit instead of relying on the presence of editDataFunction. This is more consistent with how the row actions are implemented. The table toolbar will be shown with an "edit" icon button if option `hasRowEdit` is set to be true. 
![image](https://user-images.githubusercontent.com/5167091/81402902-8f7dcc80-9132-11ea-883f-fff3f61b44a0.png)
- A new `rowEdit`-state was added to the 'activeBar' prop which previously only contained  'filter' and 'column'. When activeBar is set to `rowEdit` the other functionality that modifies the data of the table is disabled (except for row actions that stay enabled). When view.toolbar.activeBar = rowEdit the following is disabled: 
	- pagination
	- the row selection
	- the table sort
	- the table column reordering
	- table search
- Refactored the repeated code of a div with tooltip-svg-wrapper class into a component called TableToolbarSvgButton. The classname seems 'iot-tooltip-svg-wrapper' seems wrong, but I did not change it for this PR.
- A toolbar action `onShowRowEdit` was added. 	
- An `editDataFunction` was added and got the same signature of the `renderDataFunction` (ie, same 4 parameters) for consistency even though the fourth parameter was not needed for the examples in the storybook.
- Prop `preventInteraction` added to the Pagination in order to completely disable it
- Moved ActiveTableToolbarPropType to `TablePropTypes.js` to keep it DRY
- Added relevant unit tests
- Created relevant story and story info text that exemplifies the usage 

**Acceptance Test (how to verify the PR)**

1. Go to https://deploy-preview-1165--carbon-addons-iot-react.netlify.app/?path=/story/watson-iot-table--basic-table-with-full-rowedit-example
2. Verify RTL is working
3. Verify that there is a "row edit" icon in the toolbar. Click the button.
4. Make sure the table rows becomes editable, including the status columns that uses a custom data renderer function.
5. Verify that cancel/save examples works
6. Verify that undo example works

1. Go to https://deploy-preview-1165--carbon-addons-iot-react.netlify.app/?path=/story/watson-iot-table--stateful-example-with-expansion-and-column-resize
2. Verify that there is a "row edit" icon in the toolbar. Click the button.
3. This is example does not implement thet actual row edit toolbar or editable cells, but verify that pagination, row selection,  sorting, search, column reorder and filter is disabled or hidden.
